x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -252,11 +249,39 @@ public IndexMetadata getSnapshotIndexMetaData(RepositoryData repositoryData, Sna
         return imdBuilder.build();
     }
 
+    /**
+     * Dummy request key for deduplicating all remote cluster state requests via {@link #getRemoteStateDeduplicator}.
+     */
+    private static final Object RESULT_KEY = new Object();
+
+    /**
+     * Fetches the remote cluster state from the leader cluster, deduplicating the result if there is already an ongoing call
+     * to this method.
+     */
+    private void getRemoteState(ActionListener<ClusterState> listener) {
+        // We only allow a single remote cluster state requests at a time. The callbacks to the cluster state responses run on the
+        // transport thread and can safely assume they are fast enough so that this does not lead to seeing substantially outdated
+        // remote states as a result of a hot loop calling this method ever.
+        getRemoteStateDeduplicator.executeOnce(
```

Comment:
I think the use of the deduplicator risks outdated info (which I see mentioned in the comment, but not sure I follow the hot-ness argument). I think of it mainly as if the remote cluster is slow to response we risk someone having an application that:

1. Creates an index on the remote cluster (leader).
2. Invokes put-follow on the local cluster (follower).

The put-follow request could then fail due to seeing an outdated cluster state?

I think refactoring `CapacityResponseCache` will do what you want here. Seems like a utility we want to have - to only do one calculation and collapse queued requests into one, which is what `CapacityResponseCache` does.

Reply:
Hmm `CapacityResponseCache` turned out to be quite different from what we need here since it deals with a heavy but synchronous action. 

I implemented a simple solution similar to what we have for deduplicating repository data in the blob store repository now that we could extract and use for e.g. stats as well like we discussed in the past. Let me know if this is ok with you.
Did some quick benchmarking with this solution and it's also way superior in performance over the previous one since it deduplicates a lot more requests (with the first call causing all subsequent ones to queue up it works out quite nicely)